docs/TICCL-chain.1
.TH TICCL\-chain 1 "2029 mar 09"

.SH NAME
TICCL\-chain \- a Correction Candidates chaining tool.
.SH SYNOPSIS

TICCL\-chain [options] RANKED\-FILE

.SH DESCRIPTION

.B TICCL\-chain
Merge the result of a ranked file, created with
.B TICCL\-rank
into a chained file with Correction Candidates at a distance beyond
the original Levenshtein distance.

.SH OPTIONS
.B \-\-caseless
.RS
Calculate the Levenshtein (or edit) distances ignoring all case.
.RE

.B \-\-alph
alphabet
.RS
name of the alphabet file. This file must be in the format provided by
.B TICCL\-lexstat.
.RE

.B \-\-nounk
.RS
Skip Correction Candidates that are are equal except for an UNK character.
.RE

.B \-o
file
.RS
name for the output file. Default is to take the name of the input file and
add
.B .chained
at the back
.RE

.B \-V
or
.B \-\-version
.RS
Show VERSION
.RE

.B \-h
or
.B \-\-help
.RS
give usage information.
.RE

.B \-v
.RS
be verbose, repeat \-v to increase the verbosity level.
.RE

.SH BUGS
possibly

.SH AUTHORS
Ko van der Sloot lamasoftware@science.ru.nl

Martin Reynaert reynaert@uvt.nl


NEWS
ticcltools 0.6 2018-06-05
[Ko vander Sloot]
Intermediate release, with a lot of new code to handle N-grams
Also a lot of refactoring is done, for more clear and maintainable code.
This is work in progress still.

* TICCL-unk:
  - more extensive acronym detection
  - fixed artifreq problems in 'clean' punctuated words
  - added filters for 'unwanted' characters
  - added a ligature filter to convert evil ligatures
  - normalize all hyphens to a 'normal' one (-)
  - use a better definition of punctuation (unicode character class is not
    good enough to decide)

* TICCL-lexstat:
  - the 'separator' symbol should get freq=0, so it isn't counted
  - the clip value is added to the output filename

* TICCL-indexer:
  - indexer and indexerNT now produce the same output, using different
    strategies when a --foci files is used.

* TICCL-LDcalc:
  major overhaul for n-grams
  - added a ngram point column to the output (so NOT backward compatible!)
  - produce a '.short' list for short word corrections
  - produce a '.ambi' file with a list of n-grams related to short words
  - prune a lot of ngrams from the output

* TICCL-rank:
 - output is sorted now
 - honor the ngram-points from the new LDcalc. (so NOT backward compatible!)

* TICCL-chain: new module to chain ranked files

* TICCL-lexclean:
  -added a -x option for 'inverse' alphabet

* TICCL-anahash:
  - added a --list option to produce a list of words and anagram values

[Maarten van Gompel]
* added metadata file: codemeta.json

ticcltools 0.5 2018-02-19
[Ko van der Sloot]
* updated configuration. also for Mac OSX
* use of more ticcutils stuff: diacritics filter
* added a TICCL-mergelex program
* the OMP_THREAD_LIMIT environment variable was ignored sometimes
* TICCL-unk:
   - fixed a problem in artifreq handling
   - changed acronym detection (work in  progress)
   - added -o option
  TICCL-lexstat:
   - added TTR output
   - added -o option
  TICCL-indexer
  - now also handles --foci file. with some speed-up
  - added a -t option
  TICCL-LDcalc:
  - be less picky on a few wrong lines in the data
* added some tests
* when libroaring is installed we built roaring versions of some modules (experimental)
* updated man pages

ticcltools 0.4 2017-04-04
[Ko van der Sloot]
* first official release.
  - added functions to test on Word2Vec datafiles
  - refactoring and modernizing stuff all around

ticcltools 0.3 2016-01-21
* upload ready
ticcltools 0.2 2016-01-14
[Ko van der Sloot]
* repository moved to GitHub
* Travis support
* first 'distributable' version
* added TICCL-stats program
* added W2V-near and W2V-dist programs bases on new libticcl lib


[0.1] Ko vd Sloot
 started autoconfiscating


src/TICCL-chain.cxx
/*
  Copyright (c) 2006 - 2020
  CLST  - Radboud University
  ILK   - Tilburg University

  This file is part of ticcltools

  ticcltools is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 3 of the License, or
  (at your option) any later version.

  ticcltools is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, see <http://www.gnu.org/licenses/>.

  For questions and suggestions, see:
      https://github.com/LanguageMachines/ticcltools/issues
  or send mail to:
      lamasoftware (at ) science.ru.nl

*/
#include <unistd.h>
#include <set>
#include <map>
#include <functional>
#include <vector>
#include <cstdlib>
#include <string>
#include <stdexcept>
#include <iostream>
#include <fstream>
#include <cassert>
#include <cstring>
#include "config.h"
#ifdef HAVE_OPENMP
#include "omp.h"
#endif
#include "ticcutils/StringOps.h"
#include "ticcutils/CommandLine.h"
#include "ticcutils/PrettyPrint.h"
#include "ticcutils/Unicode.h"
#include "ticcl/ticcl_common.h"
#include "ticcl/word2vec.h"

using namespace std;
using namespace icu;

using TiCC::operator<<;

const bitType HonderdHash = high_five( 100 );
const bitType HonderdEenHash = high_five( 101 );
const string high_101 = TiCC::toString(HonderdEenHash);

unsigned int ld( const UnicodeString& in1,
		 const UnicodeString& in2,
		 bool caseless ){
  UnicodeString s1 = in1;
  UnicodeString s2 = in2;
  if ( caseless ){
    s1.toLower();
    s2.toLower();
  }
  return ldCompare( s1, s2 );
}

map<UChar,bitType> alphabet;

class chain_class {
public:
  chain_class(): chain_class( 0,false ){};
  chain_class( int v, bool c ): verbosity(v), caseless(c), cc_vals_present(true){};
  bool fill( const string&, bool );
  void debug_info( ostream& );
  void output( const string& );
  UnicodeString top_head( const UnicodeString& );
  void final_merge();
 private:
  map<UnicodeString,UnicodeString> heads;
  map<UnicodeString, set<UnicodeString>> table;
  map<UnicodeString, size_t > var_freq;
  map<UnicodeString, string> w_cc_conf;
  set<UnicodeString> processed;
  int verbosity;
  bool caseless;
  bool cc_vals_present;
};

UnicodeString chain_class::top_head( const UnicodeString& candidate ){
  UnicodeString result = heads[candidate];
  if ( !result.isEmpty() ){
    UnicodeString next = top_head( result );
    if ( !next.isEmpty() ){
      result = next;
    }
  }
  return result;
}

void chain_class::final_merge(){
  for ( auto& it : table ){
    if ( !it.second.empty() ){
      // for all entries that seem to be a 'head'
      UnicodeString head = top_head( it.first );
      assert( head != it.first );
      if ( !head.isEmpty() ){
	// so it has a higher head
	if ( verbosity > 3 ){
	  cerr << "merge: " << it.first << it.second << " into "
	       << head << table[head] << endl;
	}
	for ( const auto& s : it.second ){
	  table[head].insert( s );
	  heads[s] = head;
	}
	it.second.clear();
      }
    }
  }
}

UChar diff_char( const UnicodeString& in1, const UnicodeString& in2 ){
  UnicodeString s1 = in1;
  UnicodeString s2 = in2;
  s1.toLower();
  s2.toLower();
  UChar result = alphabet.begin()->first;
  for ( int i=0; i < s2.length(); ++i ){
    if ( s1.indexOf(s2[i] ) == -1 ){
      result = s2[i];
      break;
    }
  }
  return result;
}

bool chain_class::fill( const string& line, bool nounk ){
  vector<string> parts;
  TiCC::split_at( line, parts, "#", true );
  if ( parts.size() < 6
       || parts.size() > 7 ){
    return false;
  }
  else {
    if ( parts.size() == 6 ){
      cc_vals_present = false;
    }
    else if ( cc_vals_present == false ){
      cerr << "conflicting data in chained file, didn't expect cc_val entries" << endl;
      exit(EXIT_FAILURE);
    }
    UnicodeString a_word = TiCC::UnicodeFromUTF8(parts[0]); // a possibly correctable word
    if ( processed.find(a_word) != processed.end() ){
      // we have already seen this word. probably ranked with a clip >1
      // just ignore!
      //      cerr << "ignore extra entry for: " << a_word << endl;
      return true;
    }
    else {
      processed.insert(a_word);
      // so a new word with Correction Candidate
      size_t freq1 = TiCC::stringTo<size_t>(parts[1]);
      UnicodeString candidate = TiCC::UnicodeFromUTF8( parts[2] );
      // a Correction Candidate
      size_t freq2 = TiCC::stringTo<size_t>(parts[3]);
      if ( cc_vals_present ){
	string cc_val = parts[4];
	if ( nounk && cc_val == high_101 ){
	  //	  cerr << "diff?? " << a_word << " " << candidate << endl;
	  // one character difference
	  if ( candidate.length() > a_word.length() ){
	    UChar diff = diff_char( a_word, candidate );
	    //	    cerr << "diff=" << diff << endl;
	    if ( alphabet.find( diff ) == alphabet.end() ){
	      // this will not do. ignore!
	      cerr << "skip " << a_word << " --> " << candidate << endl;
	      return true;
	    }
	  }
	}
	UnicodeString key = a_word+candidate;
	w_cc_conf[key] = cc_val;
      }
      var_freq[a_word] = freq1;
      var_freq[candidate] = freq2;
      if ( verbosity > 3 ){
	cerr << endl << "word=" << a_word << " CC=" << candidate << endl;
      }
      UnicodeString head = heads[a_word];
      if ( head.isEmpty() ){
	// this word does not have a 'head' yet
	if ( verbosity > 3 ){
	  cerr << "word: " << a_word << " NOT in heads " << endl;
	}
	UnicodeString head2 = heads[candidate];
	if ( head2.isEmpty() ){
	  // the correction candidate also has no head
	  // we add it as a new head for a_word, with a table
	  heads[a_word] = candidate;
	  table[candidate].insert( a_word );
	  if ( verbosity > 3 ){
	    cerr << "candidate : " << candidate << " not in heads too." << endl;
	    cerr << "add " << candidate << " to heads[" << a_word << "]" << endl;
	    cerr << "add " << a_word << " to table of " << candidate
		 << " ==> " << table[candidate] << endl;
	  }
	}
	else {
	  // the candidate knows its head already
	  // add the word to the table of that head, and also register
	  // the head as an (intermediate) head of a_word
	  heads[a_word] = head2;
	  table[head2].insert( a_word );
	  if ( verbosity > 3 ){
	    cerr << "BUT: Candidate " << candidate << " has head: "
		 << head2 << endl;
	    cerr << "add " << a_word << " to table[" << head2 << "]"
		 << " ==> " << table[head2] << endl;
	    cerr << "AND add " << head2 << " as a head of " << a_word << endl;
	  }
	}
      }
      else {
	// the word has a head
	if ( verbosity > 3 ){
	  cerr << "word: " << a_word << " IN heads " << head << endl;
	}
	auto const tit = table.find( head );
	if ( tit != table.end() ){
	  // there MUST be some candidates registered for the head
	  if ( verbosity > 3 ){
	    cerr << "lookup " << a_word << " in " << tit->second << endl;
	  }
	  if ( tit->second.find( a_word ) == tit->second.end() ){
	    string msg = "Error: " + TiCC::UnicodeToUTF8(a_word)
	      + " has a heads entry, but no table entry!";
	    throw logic_error( msg );
	  }
	}
	else {
	  string msg = "Error: " +  TiCC::UnicodeToUTF8(a_word)
	    + " has no head entry!";
	  throw logic_error( msg );
	}
      }
    }
    if ( verbosity > 4 ){
      cerr << endl;
      debug_info( cerr );
    };
    return true;
  }
}

void chain_class::debug_info( ostream& db ){
  for ( const auto& it : heads ){
    db << "head[" << it.first << "]=" << it.second << endl;
  }
  for ( const auto& it : table ){
    db << var_freq[it.first] << " " << it.first
       << " " << it.second << endl;
  }
}

bool fillAlpha( istream& is,
		map<UChar,bitType>& alphabet,
		int clip ){
  string line;
  while ( getline( is, line ) ){
    if ( line.size() == 0 || line[0] == '#' ){
      continue;
    }
    vector<string> v;
    int n = TiCC::split_at( line, v, "\t" );
    if ( n != 3 ){
      cerr << "unsupported format for alphabet file" << endl;
      exit(EXIT_FAILURE);
    }
    int freq = TiCC::stringTo<int>( v[1] );
    if ( freq > clip || freq == 0 ){
      // freq = 0 is special, for separator
      UnicodeString v0 = TiCC::UnicodeFromUTF8( v[0] );
      bitType hash = TiCC::stringTo<bitType>( v[2] );
      alphabet[v0[0]] = hash;
    }
  }
  cout << "finished reading alphabet. (" << alphabet.size() << " characters)"
       << endl;
  return true;
}

bitType hash( const UnicodeString& s,
	      map<UChar,bitType>& alphabet ){
  UnicodeString us = s;
  us.toLower();
  bitType result = 0;
  bool multPunct = false;
  bool klets = false; //( s == "voörnaamfle" );
  for( int i=0; i < us.length(); ++i ){
    map<UChar,bitType>::const_iterator it = alphabet.find( us[i] );
    if ( it != alphabet.end() ){
      result += it->second;
      if ( klets ){
	cerr << "  CHAR, add " << UnicodeString( us[i] ) << " "
	     << it->second << " ==> " << result << endl;
      }
    }
    else {
      int8_t charT = u_charType( us[i] );
      if ( u_isspace( us[i] ) ){
	continue;
      }
      else if ( ticc_ispunct( charT ) ){
	if ( !multPunct ){
	  result += HonderdHash;
	  if ( klets ){
	    cerr << "PUNCT, add " << UnicodeString( us[i] ) << " "
		 << HonderdHash	 << " ==> " << result << endl;
	  }
	  multPunct = true;
	}
      }
      else {
	result += HonderdEenHash;
	if ( klets ){
	  cerr << "   UNK, add " << UnicodeString( us[i] ) << " "
	       << HonderdHash << " ==> " << result << endl;
	}
      }
    }
  }
  return result;
}

void chain_class::output( const string& out_file ){
  ofstream os( out_file );
  multimap<size_t, string,std::greater<size_t>> out_map;
  for ( const auto& t_it : table ){
    for ( const auto& s : t_it.second ){
      stringstream oss;
      oss << s << "#" << var_freq[s] << "#" << t_it.first
	  << "#" << var_freq[t_it.first];
      if ( cc_vals_present ){
	string val = w_cc_conf[s+t_it.first];
	if ( val.empty() ){
	  //	  cerr << "GEEN waarde voor " << s+t_it.first << endl;
	  bitType h1 = ::hash(s, alphabet );
	  bitType h2 = ::hash(t_it.first, alphabet );
	  bitType h_val;
	  if ( h1 > h2 ){
	    h_val = h1 - h2;
	  }
	  else {
	    h_val = h2 - h1;
	  }
	  //	  cerr << "h_val=" << h_val << endl;
	  w_cc_conf[s+t_it.first] = TiCC::toString(h_val);
	  //	  cerr << "nieuwe waarde voor " << s+t_it.first << "=" << w_cc_conf[s+t_it.first] << endl;
	}
	oss << "#" + w_cc_conf[s+t_it.first];
      }
      oss << "#" << ld( t_it.first, s, caseless ) << "#C";
      out_map.insert( make_pair( var_freq[t_it.first], oss.str() ) );
    }
  }
  for ( const auto& t_it : out_map ){
    os << t_it.second << endl;
  }
}

void usage( const string& name ){
  cerr << "usage: " << name << endl;
  cerr << "\t--caseless Calculate the Levensthein (or edit) distance ignoring case." << endl;
  cerr << "\t--alph <alphafile> name of the alphabet file." << endl;
  cerr << "\t--nounk Skip Correction Candidates that are are equal except for one extra UNK character." << endl;
  cerr << "\t-o <outputfile> name of the output file." << endl;
  cerr << "\t-h or --help this message." << endl;
  cerr << "\t-v be verbose, repeat to be more verbose. " << endl;
  cerr << "\t-V or --version show version. " << endl;
  exit( EXIT_FAILURE );
}

int main( int argc, char **argv ){
  TiCC::CL_Options opts;
  try {
    opts.set_short_options( "vVho:" );
    opts.set_long_options( "caseless,alph:,nounk" );
    opts.init( argc, argv );
  }
  catch( TiCC::OptionError& e ){
    cerr << e.what() << endl;
    usage( argv[0] );
    exit( EXIT_FAILURE );
  }
  string progname = opts.prog_name();
  if ( argc < 2	){
    usage( progname );
    exit(EXIT_FAILURE);
  }
  if ( opts.extract('h' ) ){
    usage( progname );
    exit(EXIT_SUCCESS);
  }
  if ( opts.extract('V' ) ){
    cerr << PACKAGE_STRING << endl;
    exit(EXIT_SUCCESS);
  }
  int verbosity = 0;
  while( opts.extract( 'v' ) ){
    ++verbosity;
  }
  bool caseless = opts.extract( "caseless" );
  bool nounk = opts.extract( "nounk" );
  string alphabet_name;
  opts.extract( "alph", alphabet_name );
  if ( alphabet_name.empty() ){
    cerr << "missing --alph option" << endl;
    exit(EXIT_FAILURE);
  }
  else {
    ifstream is( alphabet_name );
    cout << "start reading alphabet: " << alphabet_name << endl;
    fillAlpha( is, alphabet, 0 );
  }
  string out_file;
  opts.extract( 'o', out_file );

  if ( !opts.empty() ){
    cerr << "unsupported options : " << opts.toString() << endl;
    usage(progname);
    exit(EXIT_FAILURE);
  }
  vector<string> fileNames = opts.getMassOpts();
  if ( fileNames.empty() ){
    cerr << "missing an inputfile" << endl;
    exit(EXIT_FAILURE);
  }
  if ( fileNames.size() > 1 ){
    cerr << "only one inputfile may be provided." << endl;
    exit(EXIT_FAILURE);
  }
  string in_file = fileNames[0];
  if ( !TiCC::match_back( in_file, ".ranked" ) ){
    cerr << "inputfile must have extension .ranked" << endl;
    exit(EXIT_FAILURE);
  }
  if ( !out_file.empty() ){
    if ( !TiCC::match_back( out_file, ".chained" ) )
      out_file += ".chained";
  }
  else {
    out_file = in_file + ".chained";
  }
  if ( out_file == in_file ){
    cerr << "same filename for input and output!" << endl;
    exit(EXIT_FAILURE);
  }

  ifstream input( in_file );
  if ( !input ){
    cerr << "problem opening input file: " << in_file << endl;
    exit(1);
  }

  chain_class chains( verbosity, caseless );
  string line;
  while( getline( input, line ) ){
    if ( !chains.fill( line, nounk ) ){
      cerr << "invalid line: '" << line << "'" << endl;
    }
  }
  chains.final_merge();
  if ( verbosity > 0 ){
    string db_file = out_file + ".debug";
    ofstream db( db_file );
    chains.debug_info( db );
    cout << endl << "debug info stored in " << out_file << endl;
  }
  chains.output( out_file );
  cout << "results in " << out_file << endl;
}
